### PR TITLE
[Feature] Conference Calls : VideoGrid 

### PR DESCRIPTION
**Pull request by mejdoo**
## What's new in this PR?

This PR customises the videoGrid to allow more than 4 participants to show the video content from their camera.

### Testing

1. Make a build with : `dev`,`internal` or `experimental`
2. Join a group call with up to 4 participants

#### APK
[Download build #2288](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2288/artifact/build/artifact/wire-dev-PR2910-2288.apk)
[Download build #2289](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2289/artifact/build/artifact/wire-dev-PR2910-2289.apk)
[Download build #2290](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2290/artifact/build/artifact/wire-dev-PR2910-2290.apk)
[Download build #2292](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2292/artifact/build/artifact/wire-dev-PR2910-2292.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 1]
```diff
@@ -197,13 +197,31 @@ class CallingFragment extends FragmentHelper {
 
         gridViews.zipWithIndex.foreach { case (r, index) =>
           val (row, col, span) = index match {
+
```
Comment (mejdoo) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:200]: It look weird, but this how it was implemented for 4 participants.

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 2]
```diff
@@ -197,13 +197,31 @@ class CallingFragment extends FragmentHelper {
...
             case 0 if !isVideoBeingSent && gridViews.size == 2 => (0, 0, 2)
             case 0                                             => (0, 0, 1)
             case 1 if !isVideoBeingSent && gridViews.size == 2 => (1, 0, 2)
             case 1                                             => (0, 1, 1)
+
             case 2 if gridViews.size == 3                      => (1, 0, 2)
             case 2                                             => (1, 0, 1)
             case 3                                             => (1, 1, 1)
```
Comment (mejdoo) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:208]: It look weird, but this how it was implemented for 4 participants.

File: app/src/main/res/layout/fragment_calling.xml [hunk 3]
```diff
@@ -28,7 +28,7 @@
         android:id="@+id/video_grid"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
-        app:rowCount="2"
+        app:rowCount="6"
```
Comment (android10) [inline app/src/main/res/layout/fragment_calling.xml:31]: this variable looks like a hack. No meaningful name, no intention and a magic number :).

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala (hunk 2 above)
Comment (android10) [reply app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:208]: :open_mouth:

File: app/src/main/res/layout/fragment_calling.xml (hunk 3 above)
Comment (mejdoo) [reply app/src/main/res/layout/fragment_calling.xml]: Removed. It's redundant.

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala (hunk 2 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala]: I tried to make it generic to support more than 12 participants and I added an explanation about the logic of the implementation.

File: app/src/main/res/layout/fragment_calling.xml [hunk 4]
```diff
@@ -28,7 +28,6 @@
         android:id="@+id/video_grid"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
-        app:rowCount="2"
```
Comment (mejdoo) [inline app/src/main/res/layout/fragment_calling.xml:31]: No need to limit the number of rows of the grid until product team decides what's the maximum video previews to show in the screen.

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 5]
```diff
@@ -201,9 +201,13 @@ class CallingFragment extends FragmentHelper {
             case 0                                             => (0, 0, 1)
             case 1 if !isVideoBeingSent && gridViews.size == 2 => (1, 0, 2)
             case 1                                             => (0, 1, 1)
-            case 2 if gridViews.size == 3                      => (1, 0, 2)
-            case 2                                             => (1, 0, 1)
-            case 3                                             => (1, 1, 1)
+            // The max number of columns is 2 and the max number of rows is undefined
+            // if the index of the video preview is odd, display it in row n/2, column 1 , span 1
+            case n if n % 2 != 0                               => (n / 2, 1, 1)
```
Comment (johnxnguyen) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:206]: What's the "span"?

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala [hunk 6]
```diff
@@ -201,9 +201,13 @@ class CallingFragment extends FragmentHelper {
...
+            // else if the gridViews size is n+1 , display it in row n/2, column 0 , span 2
+            case n if gridViews.size == n + 1                  => (n / 2, 0, 2)
+            // else display it in row n/2, column 0 , span 1
+            case n                                             => (n / 2, 0, 1)
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:210]: Good idea about making computation of the numbers automatic, without the need for ever-expanding list of cases :)

File: app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala (hunk 5 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:206]: `GridLayout.spec(col, span, GridLayout.FILL, 1f)`
**col = from which column to start
span = size or number of columns to fill**

**Comment by mejdoo** [comment]
build again

**Pull request merged**